Akka 2.4 M2

chbatey opened the pull request:
* Removed non-permanent delete
* Upgraded c* driver
* Updated tests that relied on sender() in receiveRecover (https://github.com/akka/akka/issues/17885)

chbatey commented:
Marked as WIP until i get time to finish this off.

Okay I'll drop the java-driver back down, 2.1.7 will be released this week so we can try that later.

You're right about the contract, my bad. It would be nice to add tests for this but it would involve putting an abstraction around query execution to stub out the failure or using Stubbed Cassandra (scassandra.org) that doesn't support batches yet.

I double checked about the success case whether we need to return a Seq full of Success, they may add in a special case for this where we can return Seq(Nil) https://github.com/akka/akka/issues/17894

krasserm commented:
Sounds good. Thanks for keeping working on it, really appreciate your contributions.

chbatey commented:
Hi @krasserm I removed CassandraSnapshotStoreEndpoint and changed the CassandraJournal to first create the BoundStatements to highlight any rejections due to serialisation errors and then execute the BatchStatements preserving Failures from the previous step. My Scala foo is < Cassandra foo so very open to feedback if there is a nicer way to do this.

krasserm commented:
@chbatey your changes LGTM, I can only propose a few renamings/reformattings but that's only a matter of taste. Feel free to ignore.

    def asyncWriteMessages(messages: Seq[AtomicWrite]): Future[Seq[Try[Unit]]] = {
      val groupedStatements = messages.map(statementGroup)
      val batchStatements = groupedStatements.map({
        case Success(group: Seq[BoundStatement]) =>
          executeBatch(batch => group.foreach(batch.add)).map(_ => Success(()))
        case Failure(e) =>
          Future.successful(Failure[Unit](e))
      })
      Future.sequence(batchStatements)
    }
  
    private def statementGroup(atomicWrite: AtomicWrite): Try[Seq[BoundStatement]] = Try {
      atomicWrite.payload.flatMap { m =>
        val pnr = partitionNr(m.sequenceNr)
        val write = preparedWriteMessage.bind(m.persistenceId, pnr: JLong, m.sequenceNr: JLong, persistentToByteBuffer(m))
        if (partitionNew(m.sequenceNr)) Seq(preparedWriteHeader.bind(m.persistenceId, pnr: JLong), write) else Seq(write)
      }
    }

Please squash your commits and I'll merge. Awesome contribution, thanks a lot!

chbatey commented:
@krasserm - updated based on your suggestions and squashed. Always a pleasure to work on a nice, well tested code base.

krasserm commented:
@chbatey @bjgbeelen @raboof In order to avoid further divergence, I just rebased branch [wip-akka-2.4](https://github.com/krasserm/akka-persistence-cassandra/commits/wip-akka-2.4) onto master (incl. squashing some of your commits) and did a `push -f`. If you plan further work on wip-akka-2.4, please make a clean checkout. Sorry for the inconvenience. /cc @ktoso
